woocommerce/woocommerce
# Optimize the regeneration of the product attributes lookup table

### Changes proposed in this Pull Request:

#### What

This pull request introduces a new method for creating and updating entries in the product attributes lookup table in a more performant way. The currently existing method makes use of the standard WordPress and WooCommerce functions to retrieve all the relevant data needed for composing the lookup data (product, taxonomies, terms), the new method uses SQL queries for direct access to the relevant tables.

The performance gain is apparent especially in variable products having lots of variations. Tests on a product having about 270 variation have shown a 100x improvement in the time needed to regenerate the lookup data.

The new method can only be used if products are stored in the posts table (and product attributes are standard WordPress taxonomies). Additionally, a new setting is introduced, `woocommerce_attribute_lookup_optimized_updates`, to enable the usage of the new method; it's available in the _WooCommerce - Settings - Product - Advanced_ settings page and it's disabled by default.

![image](https://github.com/woocommerce/woocommerce/assets/937723/c7a5ab1e-e096-4196-a682-11cacaf9a94b)

[The CLI tools](https://github.com/woocommerce/woocommerce/pull/47311) are also adjusted so that by default the `regenerate` and `regenerate_for_product` commands use the new method when possible, but an optional argument is added to use the old method instead.

This pull request also changes the default value of the background table regeneration step size (the amount of products that are processed in each run of the scheduled action) from 10 to 100. Each step involves a call to `wc_get_products`, which is an expensive operation, and there's also the overhead of scheduling and running the actions; the value of 10 was too conservative and that made the regeneration process to be slower than it should. A value of 100 is still reasonable and provides a significant performance gain (additional to the other changes introduced here).

#### How

Data is (re-)created in the product attributes lookup table when any of the following happens:

1. A background table regeneration is triggered from the WooCommerce tools page or from CLI (`initiate_regeneration` command).
2. A product (including variations) is created or updated (only changes to product attributes trigger the re-creation of entries in the lookup table).
3. A synchronous table regeneration is triggered from CLI (`regenerate` and `regenerate_for_product` commands).

For 1 and 2, the new method is used only if the `woocommerce_attribute_lookup_optimized_updates` setting is set to `yes` *and* products are stored as custom post types. The later is checked as follows: `is_a( \WC_Data_Store::load( 'product' )->get_current_class_name(), 'WC_Product_Data_Store_CPT', true );`.

For 3, the new method is used by default as long as products are stored as custom post types (the value of the `woocommerce_attribute_lookup_optimized_updates` setting is irrelevant). The `--disable-db-optimization` argument can be added to force the use of the old method.

Note: updating the stock status of a product or variation will update the corresponding row(s) in the lookup table accordingly, and deleting a product or variation will of course delete the corresponding rows; but that's irrelevant in the scope of this pull request.

Closes #46699.

### How to test the changes in this Pull Request:

You need a WooCommerce install populated with products in order to test these changes, you can use [Smooth Generator](https://github.com/woocommerce/wc-smooth-generator) for that.

Also, you can run `wp wc palt info` before starting your tests, and between testing steps, to verify that the rows and products count stays consistent. As for verifying the table contents, use your favourite database querying tool or simply run `wp db query "select * from wp_wc_product_attributes_lookup where product_or_parent_id=<product id>"`.

1. Try a full table regeneration triggered from the WooCommerce tools page (WooCommerce - Status - Tools). Run the _"Regenerate the product attributes lookup table"_ tool, remove the `&action=regenerate_product_attributes_lookup_table` from the URL once the page has reloaded, and reload the page manually (or run `wp action-scheduler run`) until the button text goes from "Filling in progress" to "Regenerate" again. Try with and without the "Optimized updates" setting being active. Verify that the table updates appropriately.

2. Enable direct lookup table updates (_WooCommerce - Settings - Product - Advanced_, "Direct updates" setting). Try creating a new product, adding a new variation to an existing product, or modifying the attributes for a product. Try with and without the "Optimized updates" setting being active. Verify that the table updates appropriately.

3. Run [the CLI tools](https://github.com/woocommerce/woocommerce/pull/47311) for synchronous product and full table regeneration (`regenerate` and `regenerate_for_product`), with and without the `--disable-db-optimization` argument. Try with a variable product having a lot of variations, the process should be much faster with the optimization.

4. Although not impacted by this pull request, try that deletion and stock status changing on products and variations still update the lookup table as expected.

### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
